### PR TITLE
ComputeInstances, Datastores and HostedBy

### LINKED ISSUE #46: Networks, Software and Compute Instances
@EphyraSoftware/ephyra-dev May or may not be an issue, but just so that we have a public record of the discussion.

Continuing my work on `ComputeInstances`, I'm debating how it slots in with `Networks`

So the way we've implemented `Networks`, `ApplicationSoftware` and `SystemSoftware` can be associated with a `GroupingEntity` which is then associated with a `NetworkEntity`. 

# 1
Do we want to allow this continue, and allow the `ComputeInstance` parent relationship to be optional, and add `ComputeInstance` as just another Grouping under `GroupingEntity`? 
Thereby, if Software exists on the Network without a host, just presume the user doesn't care what it runs on, and only render specifics when a host is present? 

```
Network 
  -> Grouping
     -> Datastore
     -> *Application Software
     -> *System Software
     -> Compute Instance
          -> Application Software
          -> System Software
```

because in most instances, particularly for smaller use cases, the answer for "what does it run on?" might just be "Idk, a server?"

# 2
Or, 
should `ComputeInstance` simply be the receptacle for all `ApplicationSoftware` and `SystemSoftware`, and replace all references to software in a `GroupingEntity` with `ComputeInstances` which contain software? 

```
Network 
  -> Grouping
     -> Datastore
     -> Compute Instance
          -> Application Software
          -> System Software
```

### PR DESCRIPTION
**Pull request by Sub-Xaero**
Closes #46

### CODE REVIEW

**Comment by Sub-Xaero** [comment]
Should now be ready for all your finest critiquing.

File: acropolis-service/src/test/kotlin/org/ephyra/acropolis/service/DatastoreServiceTest.kt [hunk 1]
```diff
@@ -1,94 +1,87 @@
...
+    override fun listeners(): List<TestListener> = listOf(FixtureListener {
+        fixture = DatastoreServiceTestFixture()
+    })
 
     init {
-        "Create a new datastore" {
-            every { projectService.get(testProjectName) } returns mockk()
-            testClass.create(myStoreName, testProjectName)
-            verify { persistence.create(datastore = any()) }
+        "Create reverse proxy with base software missing, throws exception" {
```
Comment (ThetaSinner) [inline acropolis-service/src/test/kotlin/org/ephyra/acropolis/service/DatastoreServiceTest.kt]: references to reverse proxy here

File: acropolis-shell/src/main/kotlin/org/ephyra/acropolis/shell/LinkCommand.kt [hunk 2]
```diff
@@ -14,7 +15,10 @@ class LinkCommand {
     private lateinit var appState: AppState
 
     @Autowired
-    private lateinit var networkService: INetworkService
+    private lateinit var computeInstanceService: ComputeInstanceService
+
+    @Autowired
+    private lateinit var networkService: NetworkService
```
Comment (ThetaSinner) [inline acropolis-shell/src/main/kotlin/org/ephyra/acropolis/shell/LinkCommand.kt]: INetworkService

File: acropolis-shell/src/main/kotlin/org/ephyra/acropolis/shell/LinkCommand.kt [hunk 3]
```diff
@@ -14,7 +15,10 @@ class LinkCommand {
     private lateinit var appState: AppState
 
     @Autowired
-    private lateinit var networkService: INetworkService
+    private lateinit var computeInstanceService: ComputeInstanceService
```
Comment (ThetaSinner) [inline acropolis-shell/src/main/kotlin/org/ephyra/acropolis/shell/LinkCommand.kt]: IService

File: acropolis-shell/src/main/kotlin/org/ephyra/acropolis/shell/SpecializeCommand.kt [hunk 4]
```diff
@@ -2,13 +2,17 @@ package org.ephyra.acropolis.shell
...
+import org.ephyra.acropolis.service.impl.DatastoreService
 import org.ephyra.acropolis.service.impl.ReverseProxyService
 import org.springframework.beans.factory.annotation.Autowired
 import org.springframework.shell.standard.ShellComponent
 import org.springframework.shell.standard.ShellMethod
 
 @ShellComponent
 class SpecializeCommand {
+    @Autowired
+    private lateinit var datastoreService: DatastoreService
```
Comment (ThetaSinner) [inline acropolis-shell/src/main/kotlin/org/ephyra/acropolis/shell/SpecializeCommand.kt]: IService

File: acropolis-persistence/src/main/kotlin/org/ephyra/acropolis/persistence/impl/GroupingRepository.kt [hunk 5]
```diff
@@ -2,5 +2,16 @@ package org.ephyra.acropolis.persistence.impl
 
 import org.ephyra.acropolis.persistence.api.entity.GroupingEntity
 import org.springframework.data.repository.CrudRepository
+import java.util.*
```
[the remark this answers is not in the document]
Comment (Sub-Xaero) [reply acropolis-persistence/src/main/kotlin/org/ephyra/acropolis/persistence/impl/GroupingRepository.kt]: IntelliJ did it, Honest guv.

[the remark this answers is not in the document]
Comment (Sub-Xaero) [reply acropolis-persistence/src/main/kotlin/org/ephyra/acropolis/persistence/impl/GroupingRepository.kt]: Yeah, I had alreadyy coreccted that. Every time I optimise imports it puts it back.

[the remark this answers is not in the document]
Comment (ThetaSinner) [reply acropolis-persistence/src/main/kotlin/org/ephyra/acropolis/persistence/impl/GroupingRepository.kt]: Haha, yeah it does! Very few Java developers I've seen around the internet like that style but you have to turn it off in IntelliJ or spot and remove it manually

**Comment by Sub-Xaero** [comment]
Jawohl, it is done.

**Pull request merged**